CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+  }
+  for (unsigned int i = 0; i < threads.size(); i++)
+  {
+      threads[i].join();
+  }
+  // Update the output table columns.
+  mtx.lock();
+  for (unsigned int i = 0; i < numberOfThreads; i++)
+  {
+      vtkDoubleArray * bufferArray = (bufferArrays[i].Get());
+      for (unsigned int r = 0; r < bufferArray->GetNumberOfTuples(); r++)
+      {
+          double tupleValues[3] = {0.0, 0.0, 0.0};
+          tupleValues[0] = bufferArray->GetTuple3(r)[0]; // Output table row index
+          tupleValues[1] = bufferArray->GetTuple3(r)[1]; // Cross-section area
+          tupleValues[2] = bufferArray->GetTuple3(r)[2]; // CE diameter
+          crossSectionAreaArray->SetValue((int) tupleValues[0], tupleValues[1]);
+          ceDiameterArray->SetValue((int) tupleValues[0], tupleValues[2]);
+      }
+  }
+  mtx.unlock();
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+vtkCrossSectionComputeWorker::vtkCrossSectionComputeWorker()
+{
+}
+
+vtkCrossSectionComputeWorker::~vtkCrossSectionComputeWorker()
+{
+}
+
+void vtkCrossSectionComputeWorker::operator () (vtkMRMLNode * inputCenterlineNode,
+                             vtkPolyData * closedSurfacePolyData,
+                             vtkDoubleArray * bufferArray,
+                             unsigned int startPointIndex,
+                             unsigned int endPointIndex)
+{
+#if DEVTIME != 0
+    double startTime = GetTimeOfDay();
+#endif
+    for (unsigned int i = startPointIndex; i <= endPointIndex; i++)
+    {
+        // Get the contout polydata
+        vtkPolyData * polydata = ComputeCrossSectionPolydata(inputCenterlineNode, closedSurfacePolyData, i);
+        {
+            // Get the surface area and circular equivalent diameter
+            vtkNew<vtkMassProperties> crossSectionProperties;
+            crossSectionProperties->SetInputData(polydata);
+            const double crossSectionSurfaceArea = crossSectionProperties->GetSurfaceArea();
+            const double ceDiameter = (sqrt(crossSectionSurfaceArea / vtkMath::Pi())) * 2;
+            
+            bufferArray->InsertNextTuple3((double) i, crossSectionSurfaceArea, ceDiameter);
+        }
+        polydata->Delete();
+    }
+    
+    // Manage last point for an input centerline model
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        // Cross-section area is not calculated at last point for a model.
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        if (endPointIndex == (numberOfPoints - 1))
+        {
+            const unsigned int blockLastPointIndex = bufferArray->GetNumberOfTuples() - 1;
+            const unsigned int blockBeforeLastPointIndex = bufferArray->GetNumberOfTuples() - 1 - 1;
+            bufferArray->SetTuple3(blockLastPointIndex,
```

Comment:
Have you allocated memory for this tuple? If not, then it may corrupt the memory as you write it. Why don't you use InsertNextTuple3 instead?